### PR TITLE
Fix preview section numbering issue

### PR DESCRIPTION
**Pull request by rundel**
This is an attempt to fix the section numbering issue as reported in #497 as well as an additional issue with code output being interpreted as a section (due to the `##`).

I have complete removed the use of `# Placeholder` when an h3 is not found since this doesn't seem necessary (but I may have missed a use case).

In order to fix the code issue I've added an addition grep to identify code blocks and then checking for the `##` occuring inside or outside a block by counting the number of preceeding ```.

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -143,7 +143,9 @@ merge_chapters = function(files, to, before = NULL, after = NULL, orig = files)
   preview = opts$get('preview'); input = opts$get('input_rmd')
   content = unlist(mapply(files, orig, SIMPLIFY = FALSE, FUN = function(f, o) {
     x = readUTF8(f)
-    x = if (preview && !(o %in% input)) create_placeholder(x) else {
+    x = if (preview && !(o %in% input)) {
+      create_placeholder(x)
+    } else {
```
Comment (yihui) [inline R/utils.R]: In the future, please refrain from mixing cosmetic changes with essential changes in the same pull request.

File: R/utils.R [hunk 2]
```diff
@@ -162,10 +164,24 @@ create_placeholder = function(x) {
...
+  code_blocks = grep('^```', x)
+
+  # Checks for ^#{2,} potentially occuring in a code block
+  # for each match count the preceeding occurances of ^```
+  #  - odd counts = in code block
+  #  - even counts = section / subsection / etc.
+  in_block = vapply(poss_sections, function(x) as.logical(sum(x>code_blocks) %% 2), TRUE)
+
+  sections = poss_sections[!in_block]
+  h4 = x[sections]
```
Comment (yihui) [inline R/utils.R]: I don't have a strong opinion on this. I'm not against your improvement, but I don't mind if certain comments in the R code are treated as section headings, either. As I said, it is for preview only, and rough edges should be fine.

File: R/utils.R [hunk 3]
```diff
@@ -162,10 +164,24 @@ create_placeholder = function(x) {
...
+
   c(
-    '', placeholder(head(h1, 1)), placeholder(head(h2, 1)),
-    placeholder(h3[1], '# Placeholder'), '', h4
+    '',
+    placeholder(head(h1, 1)), placeholder(head(h2, 1)), placeholder(head(h3, 1)),
```
Comment (yihui) [inline R/utils.R]: So this is the actual fix to the ToC issue #497. There is indeed an edge case, for which I used `# Placeholder`, but the case is fairly minor, so I'm okay with removing the placeholder title.

File: R/utils.R (hunk 2 above)
Comment (yihui) [reply R/utils.R]: @rundel FYI: With `devtools::install_github(c('yihui/xaringan', 'rstudio/bookdown'))`, comments in fenced code blocks will no longer be treated as section headers.

Comment (rundel) [reply R/utils.R]: 👍

**Pull request closed without merging**